stdlib/public/core/KeyPathIterable.swift

```diff
@@ -0,0 +1,112 @@
+  associatedtype AllKeyPaths : Collection
+    where AllKeyPaths.Element == PartialKeyPath<Self>
+
+  /// A collection of all custom key paths of this value.
+  var allKeyPaths: AllKeyPaths { get }
+}
+
+public extension KeyPathIterable {
+  /// An array of all custom key paths of this value and any custom key paths
+  /// nested within each of what this value's key paths refers to.
+  var recursivelyAllKeyPaths: [PartialKeyPath<Self>] {
+    var result: [PartialKeyPath<Self>] = []
+    for kp in allKeyPaths {
+      result.append(kp)
+      if let nested = self[keyPath: kp] as? _KeyPathIterableBase {
+        for nkp in nested._recursivelyAllKeyPathsTypeErased {
+          result.append(kp.appending(path: nkp)!)
+        }
+      }
+    }
+    return result
+  }
+}
+
+public extension KeyPathIterable {
+  var _allKeyPathsTypeErased: [AnyKeyPath] {
+    return allKeyPaths.map { $0 as AnyKeyPath }
+  }
+  var _recursivelyAllKeyPathsTypeErased: [AnyKeyPath] {
+    return recursivelyAllKeyPaths.map { $0 as AnyKeyPath }
+  }
+}
+
+public extension KeyPathIterable {
+  /// Returns an array of all custom key paths of this value, to the specified
+  /// type.
+  func allKeyPaths<T>(to _: T.Type) -> [KeyPath<Self, T>] {
+    return allKeyPaths.compactMap { $0 as? KeyPath<Self, T> }
+  }
+
+  /// Returns an array of all custom key paths of this value and any custom key
+  /// paths nested within each of what this value's key paths refers to, to
+  /// the specified type.
+  func recursivelyAllKeyPaths<T>(to _: T.Type) -> [KeyPath<Self, T>] {
+    return recursivelyAllKeyPaths.compactMap { $0 as? KeyPath<Self, T> }
+  }
+
+  /// Returns an array of all custom writable key paths of this value, to the
+  /// specified type.
+  func allWritableKeyPaths<T>(to _: T.Type) -> [WritableKeyPath<Self, T>] {
+    return allKeyPaths(to: T.self)
+      .compactMap { $0 as? WritableKeyPath<Self, T> }
+  }
+
+  /// Returns an array of all custom writable key paths of this value and any
+  /// custom writable key paths nested within each of what this value's key
+  /// paths refers to, to the specified type.
+  func recursivelyAllWritableKeyPaths<T>(
+    to _: T.Type
+  ) -> [WritableKeyPath<Self, T>] {
+    return recursivelyAllKeyPaths(to: T.self)
+      .compactMap { $0 as? WritableKeyPath<Self, T> }
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Collection conformances
+//===----------------------------------------------------------------------===//
+
+extension Array : KeyPathIterable {
+  public typealias AllKeyPaths = [PartialKeyPath<Array>]
+  public var allKeyPaths: [PartialKeyPath<Array>] {
+    return indices.map { \Array[$0] }
+  }
+}
+
+extension Dictionary : KeyPathIterable {
+  public typealias AllKeyPaths = [PartialKeyPath<Dictionary>]
+  public var allKeyPaths: [PartialKeyPath<Dictionary>] {
+    return keys.map { \Dictionary[$0]! }
```

Comment:
Seems like this (and really all collections) should follow the array example of giving access via `indices`, yielding key paths to key-value pairs here. If you make `Dictionary.Values` key-path iterable, too, a user could write `myDictionary.values.allKeyPaths` to get this result.

Reply:
Good point. Key paths using `subscript` and `Dictionary.Index` currently causes a crasher on our branch. This may have been fixed on master. We'll make sure to update these once we merge.